### DISCUSSION TITLE
get selected text coordinates

### DISCUSSION THREAD

**Discussion by subhasispanda98** [stephanrauh/ngx-extended-pdf-viewer#1050] [Q&A]
When the user selects the text from pdf, how to get the coordinates of that selected or highlighted text? How to do that? Please help me!

**Comment by stephanrauh** [comment]
I don't understand what you're trying to say. Here's something that may help you: https://www.textbooks.com/Catalog/LDL/Basic-English-Grammar.php

**Reply by stephanrauh** [reply]
Yeah, that's an improvement. If you consider me pedantic, take in mind that this is my leisure-time project. I do it because I think it's cool to help other developers. Thing is, it eats a lot of my spare time. When I run into a question that's hardly incomprehensible I get annoyed, and I tend to close the question without answering it. There's a reason why this kind of behavior is called "to spew out a question" in German language. It annoys me, and when I'm annoyed, I don't want to go to great length to answer your question in polite words.

The other aspect is that other developers should be able to parse your question. I suppose native speakers don't have much of a problem. They're used to the lack of grammar by reading countless SMS and Whatsapp messages. But what about the non-native speakers? They don't have the context allowing them to fill in the missing gaps. I'll rephrase your question for the benefit of those developers.

**Comment by stephanrauh** [comment]
You're the third developer requesting this feature. So it seems that's a fairly common requirement. Let's open an enhancement ticket. What precisely do you need? What's your definition of "coordinates"?

**Reply by TLucasIND** [reply]
For me the feature would be very cool too. As @aysha-statvalu did not answer, here is what I'd love to see:
Adding a possibility to get the coordinates of the cursor when clicking in the PDF (maybe in the user space or in the absolute PDF-coordinates (I'd prefer latter)). 
In #972 someone asked for a similar option, but in connection with dragging a PDF to the viewer (I suppose). 
@stephanrauh: you answered that you 
> "could add the clientX and clientY coordinates to the `fileInputChange `event"

but that would be too an too exotic use-case. 
Maybe the use-case with the click event is not as exotic and you might consider it?

Maybe using the eventbus is an option too.

**Reply by stephanrauh** [reply]
I'm not sure about the PDF coordinates, but I guess it's possible to get the coordinates relative to the page. Maybe a percentage is more useful. What do you think?

**Reply by stephanrauh** [reply]
(From my point of view, the PDF viewer simply displays images, so I always wonder why so many developers consider the coordinates useful. There's no connection to the text of the PDF file. If you we to say something like "I want you to highlight this sentence" or "there's an image on the page, and I want to select it", I'd understand the use-case. Unfortunately, I simply don't know how to find a text or an image from the coordinates because pdf.js is such a complex beast. The information is all there, but it's hidden between several abstraction layers).

**Reply by stephanrauh** [reply]
After investigating the topic, I believe it's better if you implement the feature yourself. Currently, there's no such thing as an `onTextSelected` event. But it's easy to add the `select` event listener to the textlayer:

```html
<ngx-extended-pdf-viewer ... (textLayerRendered)="addSelectListenerToPage($event)">
```

```typescript
public addSelectListenerToPage(event: TextLayerRenderedEvent): void {
  event.source.textLayerDiv.addEventListener("select", (event: Event) => { whateverYouWantToDo(); });
}
```

The event doesn't contain information about the text selection. You can get this info via `document.getSelection()`.

According to https://javascript.info/selection-range#selection-events, there's also a `selectionChangeEvent` that give you more insight about what the user is doing.

I suggest you explore these options and tell me if it works. I'd also like to invite you to contribute a demo to the showcase. I don't want to implement the `(textSelected)`.event in the library, but there's nothing wrong with documenting the do-it-yourself solution on the showcase.

https://www.w3schools.com/jsref/event_onselect.asp

**Reply by TLucasIND** [reply]
In my case I know the coordinates of each text in the pdf. The customer would like to click on a word in the pdf and start another action with this word.
Currently I implemented that feature the way that the user needs to highlight the text (by selecting it via click and dragging over the text). Here I can evaluate the text via `(mouseup)`.
In contrast to that knowing the coordinates of the click would enable me to watch the corresponding text up in my table.
But that's just a draft... 
Maybe generating some kind of "links" for texts would be an option too. I did not investigate further, but I think one needs to enter an actual URL to get the pdf-viewer to interpret it as a clickable. 

I'm going to read your suggestions, thanks for that!

**Reply by TLucasIND** [reply]
Hi Stephan, I looked into your hints. Getting the selection-event is not a real problem. I had solved it like this before already:
`<ngx-extended-pdf-viewer id='pdf-viewer-component' ... (mouseup)="storeSelectedText() ..."` and with the text I did what I wanted to do.
But my current approach is to get the coordinates of the cursor over the pdf-viewer-element like so: 
`<ngx-extended-pdf-viewer id='pdf-viewer-component' ... (click)="onClick($event)"`
and then 
`onClick(event: MouseEvent) {let x = event.clientX; let y = event.clientY; ...}`
But these coordinates aren't (of course) the coordinates in the pdf.

In my Backend-Engine I have a table where every coordinate in the pdf of all significant words is stored.
When getting the right coordinates of a click, I could simply look up in my table and knew, which word was clicked on.
A drag-selection wouldn't be neccessary anymore.

I know, its kind of a hack, but I need to get rid of the drag-selection-process.

edit: had misspelled your name, sorry

**Reply by stephanrauh** [reply]
You can either add the `(onclick)` listener to the `.page` div. Or you can calculate the position relative to the .page div from the coordinates of your approach.

The problem with the .page div is you have to add the listener with standard JavaScript (as opposed to using Angular). Among other things, this means you need to activate the change detection of Angular manually.

The screenshot shows where you're currently catching the click event (upper red rectangle) and the div you're interested in (bottom-most red rectangle):

![image](https://user-images.githubusercontent.com/3045767/177711653-3703910f-e188-4a6e-a783-5bae49e74c34.png)

**Reply by TLucasIND** [reply]
Thanks for your hint. I just naively tried to call the div by class.
`var myPages = document.getElementsByClassName('page');`
`var firstPage = myPages.item(0);`
`firstPage?.addEventListener('click',(e: Event) => {console.log(e);});` 

The result is somewhat similar to the first approach.
I can find multiple pairs of coordinates. At first I thought I could work with "pageX/pageY" as they seemed to deliver the most plausible values. But when moving the inner scrolling bar of the pdf viewer, this result is "messed up" too. I think there is a way to subtract the scroll-offset (but I cant seem to understand what the "offset"-coordinates really mean here).

<img width="571" alt="Screenshot 2022-07-07 104851" src="https://user-images.githubusercontent.com/49560161/177733135-fa63c4ba-052a-4bba-8756-1aeefd21acab.png">

**Reply by stephanrauh** [reply]
Yes, indeed. Coordinates in the browser are an art of its own. Stackoverflow is full of questions and answer about this topic.

As for adding the event listener: it's a good idea to do this in the `(pageRendered)` event. Pages are drawn lazily, and the viewer also removes pages again to save memory. So it may happen that the same page is rendered more than once.
